src/mon/Monitor.cc

```diff
@@ -5337,6 +5337,27 @@ void Monitor::count_metadata(const string& field, Formatter *f)
   f->close_section();
 }
 
+bool Monitor::check_daemon_versions(map<int,std::pair<int, string>> &error){
+  int i = 0;
+  bool warn = false;
+  int start = 0;
+  string first_daemon = "Hello";
+  for (auto& p : mon_metadata) {
+    auto q = p.second.find("ceph_version");
+    string current_daemon = q->second;
+    if (start == 0){
+      first_daemon = current_daemon;
```

Comment:
Instead of returning errors which are ceph_versions that don't match the first daemon, I'd prefer if we also returned a map of versions and lists of daemons.  The caller can mention how man different version are present and for each version how many daemons are involved.  More logic is needed on how to present the information.  For example, if you had 9 OSDs and 3 different versions with 3 OSDs per version.  You can't say which are "incorrect."  You have an inconsistent state of 3 different ceph version.  

Is the first daemon going to be the leader mon?  That might justify saying the daemons have incorrect versions.

Reply:
So the _error_ in this function could be renamed _versions_ would be a map< version, string > and let the caller sort out what is inconsistent or incorrect (if there is a definitive **correct** version).

Reply:
To get a list of different versions a new table would need to be created to store them as well as a new function to allow the user to specify and add. In which case maybe the current method could be the default for when said table is empty?

With the current version the monitor does not store a ceph version so the first daemon had to be used instead, this does come with the inherent risk that the first daemon may be the only one that is incorrect, though hopefully if the use sees their intended version listed as incorrect they will realize this.

Reply:
I'd prefer the code not use the first OSD or the first daemon as the "correct" version.  It isn't clear to me that mon_metadata puts the first OSD first.  If anything during an upgrade the "correct" version is the highest version.  I'd like the case I describe above to be handled by declaring the 3 OSDs with higher version numbes as "correct" and then describe the other 6 OSDs in the health detail.  (9 OSDs with every 3 OSDs with a different version). I'd like the errors map to be renamed and the first part of the map is the version.  The second part can be the actual daemon name i.e. osd.3 mon.a.

_ceph node ls_ yields this output.  From the information below you know you have mon.a, osd.0, osd.1, osd.2, and mgr.x.
See how this is generated probably partially by _Monitor::print_nodes()_ in _mon/Monitor.cc_.  It needs hostname to parse the daemo names which in my case is bionic-laptop.

This will allow you to say "daemon osd.1 is running an inconsistent ceph version:..."

```
{
    "mon": {
        "bionic-laptop": [
            "a"
        ]
    },
    "osd": {
        "bionic-laptop": [
            0,
            1,
            2
        ]
    },
    "mgr": {
        "bionic-laptop": [
            "x"
        ]
    }
}
```

Reply:
In real life there would be multiple hosts.  I generated the output on a vstart cluster on my laptop.